mmwsts.c

```diff
@@ -0,0 +1,1069 @@
+struct stsScheme_struct {
+  char type; /* 'i' for identifier, 's' for scheme. */
+  nmbrString *scheme; /* The scheme */
+  vstring typesetting; /* The typesetting to be displayed */
+  int varCount; /* The number of variables in the scheme */
+};
+
+/* The structure containing information about the STS variable tokens */
+struct stsVar_struct {
+  long stsType; /* type of the token in the STS (must be a constant tokenId) */
+  long stsSchemeId; /* number of the schemed in which this variable is defined + 1. */
+  vstring anchor; /* anchor for the STS, or "" if not allocated yet */
+  flag isTerminal; /* whether or not this is one of the terminal type constants */
+};
+
+/* Current output format for STS */
+vstring stsFormat = "";
+
+/* STS header to be added to the HTML HEAD element of each generated page */
+vstring stsHeader = "";
+
+/* Prefix/Suffix for the output of each formula when in-text (in comments) */
+vstring stsInText[2] = { "", "" };
+
+/* Prefix/Suffix for the output of each formula when displayed (in proofs) */
+vstring stsDisplayed[2] = { "", "" };
+
+/* The pointer containing the structured typesetting replacement schemes */
+struct stsScheme_struct *stsScheme = NULL;
+long stsSchemes = 0;
+
+/* The pointer containing the structured typesetting replacement schemes */
+struct stsVar_struct *stsVar = NULL;
+long stsVars = 0;
+
+/* The basic types used for parsing tokens or comments */
+nmbrString *stsTerminalTypes = NULL_NMBRSTRING;
+
+/* debug trace levels for STS */
+flag dbs1 = 0; /* shows top-level calls to STS printing */
+flag dbs3 = 0; /* shows substitution results */
+flag dbs5 = 0; /* shows all substitutions tried */
+flag dbs7 = 0; /* shows details of substitution "unification" */
+flag dbs9 = 0; /* shows details of substring searches in "unification" */
+flag stsUseCache = 0; /* store generation results into cache */
+
+/**/
+void debugOn() {
+  dbs1 = 1; /* shows top-level calls to STS printing */
+  dbs3 = 1; /* shows substitution results */
+  dbs5 = 1; /* shows all substitutions tried */
+  dbs7 = 0; /* shows details of substitution "unification" */
+  dbs9 = 0; /* shows details of substring searches in "unification" */
+}
+
+/**/
+void debugOff() {
+  dbs1 = 0; /* shows top-level calls to STS printing */
+  dbs3 = 0; /* shows substitution results */
+  dbs5 = 0; /* shows all substitutions tried */
+  dbs7 = 0; /* shows details of substitution "unification" */
+  dbs9 = 0; /* shows details of substring searches in "unification" */
+}
+
+
+vstring stsChunkName(char c) {
+  switch(c) {
+  case 'i': return "identifier";
+  case 's': return "scheme";
+  case 'd': return "display code";
+  case 't': return "text code";
+  case 'h': return "header code";
+  case 'c': return "command line";
+  case 'u': return "terminal types";
+  default: return "unknown chunk";
+  }
+}
+
+/* Cache to speed up conversions */
+hashtable stsCache;
+
+/* Math symbol comparison for bsearch */

```

Comment:
The comment should explain the result, that is limited to -1, 0 and +1, and what is delivered when.
It seems possible that the token list contains the same key multiple times in succession.  Since this is part of a binary search, data may point somewhere in the middle of such a series.  Is it guaranteed to hold just one global element (there is a suspicious active flag in the structure, that may allow a disabled and enabled element in the same series)? Is it guaranteed there is always a global element present?  If either assumption is missed, the binary search may fail.

Clarify in comment.